sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/WaitTest.java

```diff
@@ -0,0 +1,291 @@
+    }
+
+    Instant lastProcessingTime = base;
+    for (Event<Long> event : events) {
+      Duration processingTimeDelta = new Duration(lastProcessingTime, event.processingTime);
+      if (processingTimeDelta.getMillis() > 0) {
+        stream = stream.advanceProcessingTime(processingTimeDelta);
+      }
+      lastProcessingTime = event.processingTime;
+
+      if (event.element != null) {
+        stream = stream.addElements(event.element);
+      } else {
+        stream = stream.advanceWatermarkTo(event.watermarkUpdate);
+      }
+    }
+    return p.apply(name, stream.advanceWatermarkToInfinity());
+  }
+
+  private static final AtomicReference<Instant> TEST_WAIT_MAX_MAIN_TIMESTAMP =
+      new AtomicReference<>();
+
+  @Test
+  @Category({NeedsRunner.class, UsesTestStream.class})
+  public void testWaitWithSameFixedWindows() {
+    testWaitWithParameters(
+        Duration.standardMinutes(1) /* duration */,
+        Duration.standardSeconds(15) /* lateness */,
+        20 /* numMainElements */,
+        FixedWindows.of(Duration.standardSeconds(15)),
+        20 /* numSignalElements */,
+        FixedWindows.of(Duration.standardSeconds(15)));
+  }
+
+  @Test
+  @Category({NeedsRunner.class, UsesTestStream.class})
+  public void testWaitWithDifferentFixedWindows() {
+    testWaitWithParameters(
+        Duration.standardMinutes(1) /* duration */,
+        Duration.standardSeconds(15) /* lateness */,
+        20 /* numMainElements */,
+        FixedWindows.of(Duration.standardSeconds(15)),
+        20 /* numSignalElements */,
+        FixedWindows.of(Duration.standardSeconds(7)));
+  }
+
+  @Test
+  @Category({NeedsRunner.class, UsesTestStream.class})
+  public void testWaitWithSignalInSlidingWindows() {
+    testWaitWithParameters(
+        Duration.standardMinutes(1) /* duration */,
+        Duration.standardSeconds(15) /* lateness */,
+        20 /* numMainElements */,
+        FixedWindows.of(Duration.standardSeconds(15)),
+        20 /* numSignalElements */,
+        SlidingWindows.of(Duration.standardSeconds(7)).every(Duration.standardSeconds(1)));
+  }
+
+  @Test
+  @Category({NeedsRunner.class, UsesTestStream.class})
+  public void testWaitInGlobalWindow() {
+    testWaitWithParameters(
+        Duration.standardMinutes(1) /* duration */,
+        Duration.standardSeconds(15) /* lateness */,
+        20 /* numMainElements */,
+        new GlobalWindows(),
+        20 /* numSignalElements */,
+        new GlobalWindows());
+  }
+
+  @Test
+  @Category({NeedsRunner.class, UsesTestStream.class})
+  public void testWaitWithSomeSignalWindowsEmpty() {
+    testWaitWithParameters(
+        Duration.standardMinutes(1) /* duration */,
+        Duration.standardSeconds(0) /* lateness */,
+        20 /* numMainElements */,
+        FixedWindows.of(Duration.standardSeconds(1)),
+        10 /* numSignalElements */,
+        FixedWindows.of(Duration.standardSeconds(1)));
+  }
+
+  private void testWaitWithParameters(
```

Comment:
Please document what this test ensures about the parameters provided. It might also make sense to use a `Parameterized` test suite, and with `AutoValue` having a readable test config builder is pretty painless.

Reply:
Documented - however I didn't convert it to Parameterized because I really don't like the raw object arrays it requires one to create, and I think current unit test method names are helpful.

I also did a couple of small Java 8 cleanups in the code and test :)